opentelemetry/src/logs/record.rs

```diff
@@ -29,6 +32,9 @@ pub trait LogRecord {
     /// Sets the message body of the log.
     fn set_body(&mut self, body: AnyValue);
 
+    /// Sets the trace context of the log.
+    fn set_trace_context(&mut self, trace_context: TraceContext);
```

Comment:
We wouldn’t be able to do this without making that method conditional on `#[feature = “trace”]`, since that’s where `TraceId` etc are defined.

Reply:
Do you foresee any concerns with making it conditional, as the method serves no purpose without the trace feature enabled? Alternatively, should we pass Context as a parameter, allowing the user to embed TraceContext within it? Trying to evaluating these options with the goal of keeping the API as minimal as possible.

Reply:
Conditional trait methods without a default body are a hazard, but ones with a default body are generally ok.

The issue with non-default conditional trait methods is that the crate you implement it in is not the only crate that can enable features, so you can end up with impossible builds where an unrelated dependency causes you to need to implement a method you didn't have.

Default conditional trait methods are ok, because you only opt-in to implementing them when you have something meaningful to implement. Otherwise you don't have to worry about them. The drawback of this approach compared to non-conditional trait methods with arguments that may be conditionally constructible is that the method itself becomes less reliable. If, for example, you're writing an implementation that forwards to another one there may be some methods that don't forward, because they're using the default that throws the argument away.

If you'd prefer, I can make the method itself conditional and give it an empty default body, then move `TraceContext` back into the SDK. That sets out a bit of a precedent for evolving the trait too once it's stable and anything we add to it needs to have a default body.

Reply:
Thanks for explanation. I agree conditional trait is not good to have.

> If you'd prefer, I can make the method itself conditional and give it an empty default body, then move TraceContext back into the SDK. That sets out a bit of a precedent for evolving the trait too once it's stable and anything we add to it needs to have a default body.

Yes, this would be the good approach.